## Advanced Data Science Innovation - Assignment 1


In [1]:
import pandas as pd
import numpy as np

In [2]:
df_train = pd.read_csv('../data/raw/train.csv')
df_test = pd.read_csv('../data/raw/test.csv')

In [3]:
df_train.head()

Id  GP   MIN   PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  FTA   FT%  \
0  10556  80  24.3   7.8  3.0  6.4  45.7      0.1  0.3  22.6  ...  2.9  72.1   
1   5342  75  21.8  10.5  4.2  7.9  55.1     -0.3 -1.0  34.9  ...  3.6  67.8   
2   5716  85  19.1   4.5  1.9  4.5  42.8      0.4  1.2  34.3  ...  0.6  75.7   
3  13790  63  19.1   8.2  3.5  6.7  52.5      0.3  0.8  23.7  ...  1.5  66.9   
4   5470  63  17.8   3.7  1.7  3.4  50.8      0.5  1.4  13.7  ...  0.5  54.0   

   OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0   2.2   2.0  3.8  3.2  1.1  0.2  1.6            1  
1   3.6   3.7  6.6  0.7  0.5  0.6  1.4            1  
2   0.6   1.8  2.4  0.8  0.4  0.2  0.6            1  
3   0.8   2.0  3.0  1.8  0.4  0.1  1.9            1  
4   2.4   2.7  4.9  0.4  0.4  0.6  0.7            1  

[5 rows x 21 columns]

In [4]:
df_train.shape

(8000, 21)

In [5]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Id           8000 non-null   int64  
 1   GP           8000 non-null   int64  
 2   MIN          8000 non-null   float64
 3   PTS          8000 non-null   float64
 4   FGM          8000 non-null   float64
 5   FGA          8000 non-null   float64
 6   FG%          8000 non-null   float64
 7   3P Made      8000 non-null   float64
 8   3PA          8000 non-null   float64
 9   3P%          8000 non-null   float64
 10  FTM          8000 non-null   float64
 11  FTA          8000 non-null   float64
 12  FT%          8000 non-null   float64
 13  OREB         8000 non-null   float64
 14  DREB         8000 non-null   float64
 15  REB          8000 non-null   float64
 16  AST          8000 non-null   float64
 17  STL          8000 non-null   float64
 18  BLK          8000 non-null   float64
 19  TOV   

In [6]:
df_train.describe()

Id           GP          MIN          PTS          FGM  \
count   8000.000000  8000.000000  8000.000000  8000.000000  8000.000000   
mean    6856.971000    62.777875    18.576663     7.267087     2.807037   
std     3977.447579    17.118774     8.935263     4.318732     1.693373   
min        4.000000    -8.000000     2.900000     0.800000     0.300000   
25%     3413.750000    51.000000    12.000000     4.100000     1.600000   
50%     6787.500000    63.000000    16.800000     6.300000     2.400000   
75%    10299.250000    74.000000    23.500000     9.500000     3.700000   
max    13798.000000   123.000000    73.800000    34.200000    13.100000   

               FGA          FG%      3P Made          3PA          3P%  ...  \
count  8000.000000  8000.000000  8000.000000  8000.000000  8000.000000  ...   
mean      6.231213    44.608900     0.264525     0.816563    19.583700  ...   
std       3.584559     6.155453     0.384093     1.060964    16.003155  ...   
min       0.800000    21.300000    -1.100000    -3.100000   -38.500000  ...   
25%       3.600000    40.400000     0.000000     0.100000     8.400000  ...   
50%       5.400000    44.400000     0.300000     0.800000    19.500000  ...   
75%       8.100000    48.700000     0.500000     1.500000    30.600000  ...   
max      28.900000    67.200000     1.700000     4.700000    82.100000  ...   

               FTA          FT%         OREB         DREB          REB  \
count  8000.000000  8000.000000  8000.000000  8000.000000  8000.000000   
mean      1.947787    71.365825     1.077838     2.168500     3.245300   
std       1.252352    10.430447     0.785670     1.392224     2.085154   
min       0.000000   -13.300000     0.000000     0.200000     0.300000   
25%       1.000000    65.000000     0.500000     1.100000     1.700000   
50%       1.700000    71.400000     0.900000     1.900000     2.800000   
75%       2.600000    77.500000     1.500000     2.900000     4.300000   
max      11.100000   168.900000     5.500000    11.000000    15.900000   

               AST          STL          BLK          TOV  TARGET_5Yrs  
count  8000.000000  8000.000000  8000.000000  8000.000000  8000.000000  
mean      1.624513     0.648688     0.245212     1.257762     0.833625  
std       1.355986     0.407626     0.821037     0.723270     0.372440  
min       0.000000     0.000000   -17.900000     0.100000     0.000000  
25%       0.700000     0.300000     0.100000     0.700000     1.000000  
50%       1.300000     0.600000     0.200000     1.100000     1.000000  
75%       2.200000     0.900000     0.400000     1.600000     1.000000  
max      12.800000     3.600000    18.900000     5.300000     1.000000  

[8 rows x 21 columns]

In [7]:
df_test.describe()

Id           GP          MIN          PTS          FGM  \
count   3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean    7010.614109    62.853909    18.650224     7.328034     2.835404   
std     3954.173641    17.151740     8.727259     4.294724     1.688427   
min        1.000000     6.000000     3.700000     0.700000     0.300000   
25%     3644.000000    51.000000    12.200000     4.200000     1.600000   
50%     7062.000000    63.000000    17.000000     6.400000     2.500000   
75%    10402.500000    74.000000    23.300000     9.400000     3.700000   
max    13792.000000   126.000000    68.000000    33.000000    13.400000   

               FGA          FG%      3P Made          3PA          3P%  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      6.302580    44.599079     0.255962     0.796920    19.234746   
std       3.579221     6.040168     0.380987     1.052862    15.968989   
min       0.800000    25.100000    -1.000000    -2.700000   -38.000000   
25%       3.700000    40.500000     0.000000     0.100000     8.500000   
50%       5.500000    44.600000     0.300000     0.800000    19.400000   
75%       8.100000    48.500000     0.500000     1.500000    30.250000   
max      26.200000    74.600000     1.600000     4.300000    73.800000   

               FTM          FTA          FT%         OREB         DREB  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      1.399842     1.953567    71.612924     1.096025     2.179495   
std       0.926140     1.250376    10.457336     0.785678     1.371935   
min       0.000000     0.000000    23.700000     0.000000     0.200000   
25%       0.700000     1.000000    65.000000     0.500000     1.200000   
50%       1.200000     1.700000    71.500000     0.900000     1.900000   
75%       1.900000     2.600000    78.000000     1.500000     2.900000   
max       7.800000     9.800000   127.100000     6.900000    12.000000   

               REB          AST          STL          BLK          TOV  
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000  
mean      3.275783     1.636483     0.653593     0.257726     1.257910  
std       2.070646     1.335496     0.410573     0.639660     0.712449  
min       0.300000     0.000000     0.000000    -7.100000     0.100000  
25%       1.800000     0.600000     0.400000     0.100000     0.700000  
50%       2.800000     1.300000     0.600000     0.200000     1.100000  
75%       4.300000     2.300000     0.900000     0.400000     1.600000  
max      18.500000     9.000000     2.700000    14.800000     5.200000

## 3. Prepare Data

In [8]:
# Create a copy of df and save it into a variable called df_cleaned
df_cleaned = df_train.copy()

In [10]:
# Drop the column `url`
# df_cleaned.drop('url', axis=1, inplace=True)

**[3.3]** Remove leading and trailing space from the column names

In [11]:
df_cleaned.columns = df_cleaned.columns.str.strip()

**[3.4]** Extract the column `TARGET_5Yrs` and save it into variable called `target`

In [13]:
# Solution:
target = df_cleaned.pop('TARGET_5Yrs')

**[3.5]** Import StandardScaler from sklearn.preprocessing<br>
**[3.6]** Instantiate the StandardScaler<br>
**[3.7]** Fit and apply the scaling on df_cleaned

In [14]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_cleaned = scaler.fit_transform(df_cleaned)

**[3.8]** Import dump from joblib
**[3.9]** Save the scaler into the folder `models` and call the file `scaler.joblib`

In [15]:
from joblib import dump
dump(scaler, '../models/scaler.joblib')

['../models/scaler.joblib']

**[3.10]** Import train_test_split from sklearn.model_selection 

**[3.11]** Split randomly the dataset with random_state=8 into 2 different sets: data (80%) and test (20%)

**[3.12]** Split the remaining data (80%) randomly with random_state=8 into 2 different sets: training (80%) and validation (20%)

In [16]:
from sklearn.model_selection import train_test_split

# Split randomly the dataset with random_state=8 into 2 different sets: data (80%) and test (20%)
X_data, X_test, y_data, y_test = train_test_split (df_cleaned, target, test_size=0.2, random_state=8)

# Split the remaining data (80%) randomly with random_state=8 into 2 different sets: training (80%) and validation (20%)
X_train, X_val, y_train, y_val = train_test_split(X_data, y_data, test_size=0.2, random_state=8)

**[3.13]** Save the different sets in the folder `data/processed`

In [17]:
np.save('../data/processed/X_train', X_train)
np.save('../data/processed/X_val',   X_val)
np.save('../data/processed/X_test',  X_test)
np.save('../data/processed/y_train', y_train)
np.save('../data/processed/y_val',   y_val)
np.save('../data/processed/y_test',  y_test)

## 4. Get Baseline Model

**[4.1]** Calculate the average of the target variable for the training set and save it into a variable called `y_mean`

**[4.2]** Create a numpy array called `y_base` of dimensions (len(y_train), 1) filled with this value

In [18]:
# Calculate the average of the target variable for the training set
y_mean = y_train.mean()

# Create a numpy array called `y_base` of dimensions (len(y_train), 1) filled with this value
y_base = np.full((len(y_train), 1), y_mean)

**[4.3]** Import the MSE and MAE metrics from sklearn

**[4.4]** Display the RMSE and MAE scores of this baseline model

In [19]:
# Import the MSE and MAE metrics from sklearn
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import mean_absolute_error as mae

# Display the RMSE and MAE scores of this baseline model
print(mse(y_train, y_base, squared=False))
print(mae(y_train, y_base))

0.3692658517749907
0.27271453857421873


## 5. Train ElasticNet model

**[5.1]** Import the ElasticNet module from sklearn

**[5.2]** Task: instantiate the ElasticNet class into a variable called reg

**[5.3]** Task: Fit the model with the prepared data

In [21]:
# Solution:
from sklearn.linear_model import ElasticNet 

# Solution
reg = ElasticNet()

# Solution
reg.fit(X_train, y_train)

ElasticNet(alpha=1.0, copy_X=True, fit_intercept=True, l1_ratio=0.5,
           max_iter=1000, normalize=False, positive=False, precompute=False,
           random_state=None, selection='cyclic', tol=0.0001, warm_start=False)

**[5.4]** Save the fitted model into the folder `models` as a file called `elasticnet_default.joblib`

**[5.5]** Save the predictions from this model for the training and validation sets into 2 variables called `y_train_preds` and `y_val_preds`

In [25]:
# Solution:
dump(reg,  '../models/elasticnet_default.joblib')

# Solution:
y_train_preds = reg.predict(X_train)
y_val_preds = reg.predict(X_val)

**[5.6]** Display the RMSE and MAE scores of this model on the training set

In [30]:
# RMSE and MAE scores of this model on the training set
print(mse(y_train, y_train_preds, squared=False))
print(mae(y_train, y_train_preds))

0.3593035557949731
0.2619093228941328


**[5.7]** Display the RMSE and MAE scores of this model on the validation set

In [29]:
# RMSE and MAE scores of this model on the validation set
print(mse(y_val, y_val_preds, squared=False))
print(mae(y_val, y_val_preds))

0.3804573919933454
0.2770110091151148


## 6.   Push changes to GitHub

**[6.1]** Add changes to git staging area

**[6.2]** Create the snapshot of your repository and add a description

**[6.3]** Push your snapshot to Github

In [ ]:
# Add changes to git staging area
git add .

# Create the snapshot of your repository and add a description
git commit -m "first commit"

# Push your snapshot to Github
git push https://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@github.com/CazMayhem/adv_dsi_AT1.git
    

**[6.4]** Close Jupyter Lab with control (command) + c